secret-rotator/rotator/rotator_test.go

```diff
@@ -0,0 +1,425 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rotator
+
+import (
+	"bytes"
+	"reflect"
+	"sigs.k8s.io/k8s-gsm-tools/secret-rotator/config"
+	"sigs.k8s.io/k8s-gsm-tools/secret-rotator/svckey"
+	"sigs.k8s.io/k8s-gsm-tools/secret-rotator/tests"
+	"testing"
+	"time"
+
+	secretmanagerpb "google.golang.org/genproto/googleapis/cloud/secretmanager/v1"
+)
+
+var str2Time = func(str string) time.Time {
+	t, _ := time.Parse(time.RFC3339, str)
+	return t
+}
+var str2Duration = func(str string) time.Duration {
+	d, _ := time.ParseDuration(str)
+	return d
+}
+
+func TestRefresh(t *testing.T) {
+
+	newSecretId := "new-key-id"
+	newSecretValue := []byte("new-service-account-key")
+
+	// prepare provisioners for all supported types of secrets
+	provisioners := map[string]SecretProvisioner{}
+	provisioners[svckey.ServiceAccountKeySpec{}.Type()] = &tests.MockSvcProvisioner{
+		NewSecretID:    newSecretId,
+		NewSecretValue: newSecretValue,
+	}
+
+	rotator := &SecretRotator{
+		Provisioners: provisioners,
+	}
+
+	var testcases = []struct {
+		name           string
+		client         *tests.MockClient
+		spec           config.RotatedSecretSpec
+		now            time.Time
+		refresh        bool
+		expectedLabels map[string]string
+		expectErr      bool
+	}{
+		{
+			name: "Within refresh interval. Should not refresh secret.",
+
+			client: &tests.MockClient{
+				Secrets: map[string]map[string]*tests.Secret{
+					"project-1": map[string]*tests.Secret{
+						"secret-1": &tests.Secret{
+							Versions: map[string]*tests.Version{
+								"1": &tests.Version{
+									CreateTime: str2Time("2000-01-01T00:00:00+00:00"),
+									Data:       []byte("secret-data-1"),
+									State:      1, //SecretVersion_ENABLED
+								},
+							},
+							Labels: map[string]string{
+								"project":         "project-1",
+								"service-account": "service-foo",
```

Comment:
Thinking about this more, maybe these don't make sense to store as label.  You don't have test cases or code written yet to decide how to handle the case where the rotatedsecret's spec for these differs from what's already stored in the secret.  I guess I would expect them to always be updated to whatever the rotated secret has in its spec, at which point I wonder why these are here at all.  It's informative for a human manually looking at labels on this secret, but maybe not necessary?

Reply:
discussed offline. will take the config as reference, but also update the Secret Manager labels within each run.